test/Utils-spec.jsx

```diff
@@ -229,26 +217,237 @@ describe('Utils', () => {
+      nodesMatchTwoWays({ id: null }, { id: undefined });
+
+      nodesMatchTwoWays({ id: undefined }, {});
+      nodesMatchTwoWays({ id: undefined }, { id: null });
+      nodesMatchTwoWays({ id: undefined }, { id: undefined });
     });
 
+    it('should not match empty elements of different type, not distinguishing null/undefined/absent', () => {
+      nodesDoNotMatchTwoWays({}, {}, 'div', 'nav');
+      nodesDoNotMatchTwoWays({}, { id: null }, 'div', 'nav');
+      nodesDoNotMatchTwoWays({}, { id: undefined }, 'div', 'nav');
+
+      nodesDoNotMatchTwoWays({ id: null }, {}, 'div', 'nav');
+      nodesDoNotMatchTwoWays({ id: null }, { id: null }, 'div', 'nav');
+      nodesDoNotMatchTwoWays({ id: null }, { id: undefined }, 'div', 'nav');
+
+      nodesDoNotMatchTwoWays({ id: undefined }, {}, 'div', 'nav');
+      nodesDoNotMatchTwoWays({ id: undefined }, { id: null }, 'div', 'nav');
+      nodesDoNotMatchTwoWays({ id: undefined }, { id: undefined }, 'div', 'nav');
+    });
+
+    it('should match basic prop types', () => {
+      nodesMatchTwoWays({ className: 'foo' }, { className: 'foo' });
+      nodesMatchTwoWays({ id: 'foo', className: 'bar' }, { id: 'foo', className: 'bar' });
+      nodesDoNotMatchTwoWays({ id: 'foo', className: 'bar' }, { id: 'foo', className: 'baz' });
+    });
+
+    it('should check children as well, not distinguishing null/undefined/absent', () => {
+      expect(nodeMatches(
+        <div>
+          <div />
+        </div>,
+        <div />,
+      )).to.equal(false);
+
+      expect(nodeMatches(
+        <div><div /></div>,
+        <div><div /></div>,
+      )).to.equal(true);
+
+      expect(nodeMatches(
+        <div><div id={null} /></div>,
+        <div><div /></div>,
+      )).to.equal(true);
+      expect(nodeMatches(
+        <div><div /></div>,
+        <div><div id={null} /></div>,
+      )).to.equal(true);
+
+      expect(nodeMatches(
+        <div><div id={undefined} /></div>,
+        <div><div /></div>,
+      )).to.equal(true);
+      expect(nodeMatches(
+        <div><div /></div>,
+        <div><div id={undefined} /></div>,
+      )).to.equal(true);
+
+      expect(nodeMatches(
+        <div><div id={undefined} /></div>,
+        <div><div id={null} /></div>,
+      )).to.equal(true);
+      expect(nodeMatches(
+        <div><div id={null} /></div>,
+        <div><div id={undefined} /></div>,
+      )).to.equal(true);
+
+      expect(nodeMatches(
+        <div>
+          <div className="foo" />
+        </div>,
+        <div>
+          <div className="foo" />
+        </div>,
+      )).to.equal(true);
+
+      expect(nodeMatches(
+        <div>
+          <div className="foo" />
+        </div>,
+        <div>
+          <div />
+        </div>,
+      )).to.equal(false);
+    });
+
+    it('should test deepEquality with object props', () => {
+      expect(nodeMatches(
+        <div foo={{ a: 1, b: 2 }} />,
+        <div foo={{ a: 1, b: 2 }} />,
+      )).to.equal(true);
```

Comment:
I copied the cases directly from the nodeEqual ones - I'm happy to add nested cases to both but I don't think that needs to go in this PR necessarily.